src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,395 @@
+							  char *shardSplitMode);
+void ExecuteSplitBackgroundJob(int64 jobId, ShardInfo shardinfo, List *splitPoints,
+							   char *shardSplitMode);
+int64 ExecuteAvgHashQuery(ShardInfo shardinfo);
+List * FindShardSplitPoints(ShardInfo shardinfo);
+void ScheduleShardSplit(ShardInfo shardinfo, char *shardSplitMode);
+
+/*
+ * It throws an error if a concurrent automatic shard split or Rebalance operation is happening.
+ */
+void
+ErrorOnConcurrentOperation()
+{
+	int64 jobId = 0;
+	if (HasNonTerminalJobOfType("rebalance", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("A rebalance is already running as job %ld", jobId),
+					errdetail("A rebalance was already scheduled as background job"),
+					errhint("To monitor progress, run: SELECT * FROM "
+							"citus_rebalance_status();")));
+	}
+	if (HasNonTerminalJobOfType("Automatic Shard Split", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("An automatic shard split is already running as job %ld",
+						   jobId),
+					errdetail(
+						"An automatic shard split was already scheduled as background job")));
+	}
+}
+
+
+/*
+ * For a given SplitPoints , it creates the SQL query for the Shard Splitting
+ */
+StringInfo
+GetShardSplitQuery(ShardInfo shardinfo, List *splitPoints, char *shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+
+	int64 length = list_length(splitPoints);
+	appendStringInfo(splitQuery, "SELECT citus_split_shard_by_split_points(%ld, ARRAY[",
+					 shardinfo->shardId);
+
+	int32 splitpoint = 0;
+	int index = 0;
+	foreach_int(splitpoint, splitPoints)
+	{
+		appendStringInfo(splitQuery, "'%d'", splitpoint);
+
+		if (index < length - 1)
+		{
+			appendStringInfoString(splitQuery, ",");
+		}
+
+		index++;
+	}
+
+/*All the shards after the split will be belonging to the same node */
+	appendStringInfo(splitQuery, "], ARRAY[");
+
+	for (int i = 0; i < length; i++)
+	{
+		appendStringInfo(splitQuery, "%d,", shardinfo->nodeId);
+	}
+	appendStringInfo(splitQuery, "%d], %s)", shardinfo->nodeId, quote_literal_cstr(
+						 shardSplitMode));
+
+	return splitQuery;
+}
+
+
+/*
+ * It creates a background job for citus_split_shard_by_split_points and executes it in background.
+ */
+void
+ExecuteSplitBackgroundJob(int64 jobId, ShardInfo shardinfo, List *splitPoints,
+						  char *shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+	splitQuery = GetShardSplitQuery(shardinfo, splitPoints, shardSplitMode);
+	/* ereport(LOG, (errmsg(splitQuery->data))); */
+	int32 nodesInvolved[] = { shardinfo->nodeId };
+	Oid superUserId = CitusExtensionOwner();
+	BackgroundTask *task = ScheduleBackgroundTask(jobId, superUserId, splitQuery->data, 0,
+												  NULL, 1, nodesInvolved);
+}
+
+
+/*
+ * It executes a query to find the average hash value in a shard considering rows with a limit of 10GB .
+ * If there exists a hash value it is returned otherwise shardminvalue-1 is returned.
+ */
+int64
+ExecuteAvgHashQuery(ShardInfo shardinfo)
```

Comment:
may be reanme to ExecuteAvgHashQuery => ExecuteAverageHashQuery?

Reply:
renamed